hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/ToOneAttributeMapping.java

```diff
@@ -998,6 +998,22 @@ public EntityFetch generateFetch(
 				|| fetchParent.getNavigablePath() instanceof TreatedNavigablePath
 				&& parentNavigablePath.equals( fetchParent.getNavigablePath().getRealParent() );
 
+		if ( cardinality == Cardinality.ONE_TO_ONE && bidirectionalAttributeName != null && fetchParent.getReferencedMappingType() == getEntityMappingType() ) {
+			return new EntityFetchSelectImpl(
+					fetchParent,
+					this,
+					fetchablePath,
+					foreignKeyDescriptor.createDomainResult(
+							fetchablePath,
+							parentTableGroup,
+							ForeignKeyDescriptor.Nature.TARGET,
+							creationState
+					),
+					true,
+					creationState
+			);
+		}
+
 		if ( fetchTiming == FetchTiming.IMMEDIATE && selected ) {
```

Comment:
Maybe it's better to exclude the above case to entering this if-branch instead as the below code block should handle delayed/select fetching?